Adding Multichip Microbenchmarks

SeanNijjar opened the pull request:
- Bidirectional bandwidth microbenchmarks
   - This benchmark replays buffers from local erisc over the ethernet link. No data integrity check (since we don't have overheads of receiving data from noc)
   - Peak acheived was ~21.3GBps for best configuration
   - No IRAM support enabled yet

```
TT_METAL_DEVICE_PROFILER=1 {os.environ['TT_METAL_HOME']}/build/test/tt_metal/perf_microbenchmark/ethernet/test_ethernet_bidirectional_bandwidth_no_edm {num_samples}  {sample_size}  {num_channels}
```  

Latency benchmark:
```
TT_METAL_DEVICE_PROFILER=1  ./build/test/tt_metal/perf_microbenchmark/test_ethernet_hop_latencies_no_edm  {len(sample_counts)} {sample_counts_str}    {len(page_sizes)} {page_sizes_str}   {len(channel_counts)} 
 {channel_counts_str}   {len(hop_counts)} {hop_counts_str}
```


Initial Latency Results:
https://docs.google.com/spreadsheets/d/16gRDxhMGRUYQbLZeGcjvmWUY5onnmWM-b3Zfa4jJCX8/edit?usp=sharing
<img width="877" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/1865303/77196c1e-a5b4-4377-a954-32cb79a9bf34">

<img width="877" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/1865303/55cf3875-b0ed-45df-b0cf-5042840c83db">

<img width="660" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/1865303/95720556-1ae6-456b-9aed-507e9207cbd7">

<img width="605" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/1865303/0bc7a908-47df-416e-bfd0-508bbc8a6184">

SeanNijjar commented:
workflows:
post-commit: https://github.com/tenstorrent/tt-metal/tree/snijjar/ccl-microbenchmarks
multi-device unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/8714494939
microbench: https://github.com/tenstorrent/tt-metal/actions/runs/8714480105

tt-rkim commented on tests/tt_eager/module.mk:
```diff
@@ -29,7 +29,7 @@ TT_EAGER_TESTS += \
 
 TT_EAGER_TESTS_INCLUDES = $(TEST_INCLUDES) $(TT_EAGER_INCLUDES)
-TT_EAGER_TESTS_LDFLAGS = $(TT_METAL_TESTS_LDFLAGS) $(TT_LIB_LDFLAGS) -lgtest -lgtest_main
+TT_EAGER_TESTS_LDFLAGS = $(TT_METAL_TESTS_LDFLAGS) $(TT_LIB_LDFLAGS) -lgtest -lgtest_main -ltt_lib_csrc
```
Why is this needed?

tt-rkim commented on tests/scripts/run_tests.sh:
```diff
@@ -120,6 +120,7 @@ run_frequent_multi_device_pipeline_tests() {
     # Switch to modules only soon
     # run_module_tests "$tt_arch" "llrt" "$pipeline_type"
     ./tests/scripts/run_frequent_regressions_multi_device.sh
+    pytest -svv tests/tt_metal/microbenchmarks -k test_multichip_hop_latency
```
Could you please split off the specific microbenchmark into a single pytest file (such as keeping them in one place + calling only one function in a separate file)?

We're trying to redo how tests are run on the infra side, and we think we're going to go with the approach of targeting files and markers. No more `-k`

SeanNijjar replied:
I split them up so now it's a py file for each for the latency and bandwidth benchmarks. I put them in `tests/scripts/run_frequent_regressions_multi_device.sh`  because they (well mainly the latency test) needs t3000.

tt-rkim commented on tests/tt_metal/tt_metal/test_kernels/dataflow/unit_tests/erisc/test_multichip_latency.cpp:
```diff
@@ -0,0 +1,91 @@
+// // SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Delete this file please

tt-rkim commented on tests/tt_metal/tt_metal/perf_microbenchmark/ethernet/test_ethernet_hop_latencies_no_edm.cpp:
```diff
@@ -0,0 +1,537 @@
+                        receiver_kernel_ids,
+                        sender_kernel_ids
+                    );
+                    // tt::tt_metal::run_roundtrip_latency_test(
```
Is this call needed for debug?

tt-rkim commented on tests/tt_metal/tt_metal/perf_microbenchmark/ethernet/test_ethernet_bidirectional_bandwidth_no_edm.cpp:
```diff
@@ -0,0 +1,387 @@
+    std::cout << "eth_x: " << device_1->ethernet_core_from_logical_core(eth_receiver_core).x << ", " <<
+                 "eth_y: " << device_1->ethernet_core_from_logical_core(eth_receiver_core).y << "" << std::endl;
+    // bool success = false;
+    // try {
```
Are any of the commented out calls needed for debug?

tt-rkim commented on tests/scripts/run_frequent_regressions_multi_device.sh:
```diff
@@ -16,6 +16,10 @@ cd $TT_METAL_HOME
 
 pytest tests/ttnn/unit_tests/test_multi_device.py
+
+pytest tests/tt_metal/microbenchmarks/ethernet/test_ethernet_bidirectional_bandwidth_microbenchmark.py -k test_bidirectional_erisc_bandwidth
```
Hmm still looks like there's a -k here? Or are you still working on splitting it up?

SeanNijjar replied:
actually the -k is completely redundant now... I'll get rid of it :)

SeanNijjar committed:
#7528: add a multichip ring latency ubench test

Only supports T3000.

The test supports multi-channel, various hop counts (4, 8, 12), page
size, and sample count.

Each sample sends num_channels worth of payloads around the ring in
parallel. Individual samples are non-overlapping.

SeanNijjar committed:
#7528: add a bidirectional ethernet bandwidth ubench test

Adds a minimal implementation of a multi-channel, bidirectional ethernet
bandwidth benchmark. Best speeds recorded so far (without erisc IRAM
enabled) was 20-21 GBps. This is still noticeably short of the
theoretical max of the link or even the unidirectional bandwidth, when
doubled.

Additional measurements should be made after IRAM support is added
